File: Total Cost to Hire K Workers/total-cost-to-hire-k-workers.py
class Solution:
    def totalCost(self, costs, k, candidates):
        l, r = [], []
        i, j = 0, len(costs) - 1
        ans = 0
        
        while k > 0:
            while len(l) < candidates and i <= j:
                heapq.heappush(l, costs[i])
                i += 1
            while len(r) < candidates and j >= i:
                heapq.heappush(r, costs[j])
                j -= 1
            
            left = l[0] if l else float('inf')
            right = r[0] if r else float('inf')
            
            if left <= right:
                ans += left
                heapq.heappop(l)
            else:
                ans += right
                heapq.heappop(r)
                
            k -= 1
        
        return ans

File: Total Cost to Hire K Workers/total-cost-to-hire-k-workers.cpp
class Solution {
public:
    long long totalCost(vector<int>& costs, int k, int candidates) {
        priority_queue<int, vector<int>, greater<int>> l, r;
        int i = 0, j = costs.size()-1;
        long long ans = 0;
        while (k > 0) {
            while (l.size() < candidates && i <= j) {
                l.push(costs[i]);
                i++;
            }
            while (r.size() < candidates && j >= i) {
                r.push(costs[j]);
                j--;
            }
            int left = l.size() > 0 ? l.top() : INT_MAX;
            int right = r.size() > 0 ? r.top() : INT_MAX;
            if (left <= right) {
                ans += left;
                l.pop();
            } else {
                ans += right;
                r.pop();
            }
            k--;
        }
        return ans;
    }
};